compiler/src/iree/compiler/Codegen/LLVMGPU/Analysis/ROCDLThreadUniformAnalysis.cpp

```diff
@@ -0,0 +1,259 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Analysis/ROCDLThreadUniformAnalysis.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/TensorExt/IR/TensorExtOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+#define DEBUG_TYPE "rocdl-thread-uniform-analysis"
+#define LDBGS(X)                                                               \
+  LLVM_DEBUG((llvm::dbgs() << "[" << DEBUG_TYPE << "]") << X << "\n")
+
+namespace mlir::iree_compiler::ROCDL {
+//===----------------------------------------------------------------------===//
+// ThreadUniform
+//===----------------------------------------------------------------------===//
+
+ThreadUniform ThreadUniform::join(const ThreadUniform &lhs,
+                                  const ThreadUniform &rhs) {
+  if (lhs.state == dependent || rhs.state == dependent)
+    return getDependent();
+  return (lhs.state == uniform || rhs.state == uniform) ? getUniform()
+                                                        : ThreadUniform();
+}
+
+void ThreadUniform::print(llvm::raw_ostream &s) const {
+  s << (state == uninitialized ? "uninitialized"
+                               : (state == uniform ? "uniform" : "dependent"));
+}
+
+//===----------------------------------------------------------------------===//
+// ThreadUniformAnalysis
+//===----------------------------------------------------------------------===//
+/// This function is an overly conservative estimate ops that are safe to assume
+/// to be uniform. TODO: Encode this in an interface.
+static bool maybeDefinitelyWorkgroupUniform(Operation *op) {
+  if (op->hasTrait<OpTrait::ConstantLike>() ||
+      op->hasTrait<OpTrait::IsTerminator>())
+    return true;
+  if (isa<IREE::HAL::InterfaceBindingSubspanOp,
+          IREE::HAL::InterfaceConstantLoadOp,
+          IREE::TensorExt::DispatchTensorLoadOp, IREE::Util::AssumeIntOp,
+          IREE::TensorExt::DispatchWorkloadOrdinalOp, LoopLikeOpInterface>(op))
+    return true;
+  if (isa<affine::AffineDialect>(op->getDialect()))
+    return !isa<affine::AffineDmaStartOp, affine::AffineDmaWaitOp>(op);
+  return isa<arith::ArithDialect>(op->getDialect()) &&
```

Comment:
This is a rather weird test - what's this for? And why are you checking types?

At a high level - a pure function of uniform values is uniform

Reply:
I mean, we have to work this out before bufferization so we can bufferize correctly, but there's a reason I had the pass _right_ before bufferization.

Reply:
Ok, what do you mean by shaped values that are thread dependent?

Like, what's an example what you're worried about

Reply:
Yes, this is before bufferization in order to not have to update address spaces

Reply:
> Ok, what do you mean by shaped values that are thread dependent?
> Like, what's an example what you're worried about

Any vector value that has yet to be distributed. By definition, it would not be uniform, take an addf that gets distributed. 
Again, if the analysis were to be modified to be run after vector distribute, then yes, pure of uniform is uniform.

> Yes, this is before bufferization in order to not have to update address spaces

I'm claiming we should the difficult approach, to always be accurate.

Reply:
Agreed, but since we're running pre-tiling, we might want to be pessimistic

... Then again, extract_slice of a uniform tensor with uniform sizes, stride, and offset is uniform, so I see what you're getting at

I think the fact that undistributed vectors look uniform is fine actually - because they _are_ uniform ... just not real